### PR TITLE
Gutsy Gamblers

### PR DESCRIPTION
**Pull request by biskette**
PR'ing like a cat.

### CODE REVIEW

File: gutsy-gamblers/main.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            midnight = datetime.now() + timedelta(days=1)
+            # Good thing flake8 isn't 78 chars huh?
+            self.dial.midnight_delta = (datetime(year=midnight.year,
+                                                 month=midnight.month,
+                                                 day=midnight.day,
+                                                 hour=0,
+                                                 minute=0,
+                                                 second=0) - datetime.now()).seconds
+            self.dial.redraw()
+            # self.parent.speedy_time = False
```
Comment (lemonsaurus) [inline gutsy-gamblers/main.py]: It would've been better not to leave commented out code in your submission.

File: gutsy-gamblers/main.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    Button for updating location.
+    """
+    settingsscreen = ObjectProperty()
+
+    latlon = ConfigParserProperty(
+        '', 'global', datahelpers.LOCATION_LATLON, 'app', val_type=str)
+    friendlyname = ConfigParserProperty(
+        '', 'global', datahelpers.LOCATION_FRIENDLY, 'app', val_type=str)
+
+    latlong_regex = re.compile(r'^(\-?\d+(\.\d+)?),\s*(\-?\d+(\.\d+)?)$')
```
Comment (lemonsaurus) [inline gutsy-gamblers/main.py]: complicated regular expressions should be broken up into multiple lines with inline comments explaining what the different parts do for improved readability.

File: gutsy-gamblers/main.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        self.location_field.text = self.config_friendlyname
+
+    def on_config_latlon(self, instance, value):
+        """Assumes this will only ever be set on a valid key"""
+        self.close_button.disabled = False
+
+
+class SunClock(App):
+    """
+    Core class.
```
Comment (lemonsaurus) [inline gutsy-gamblers/main.py]: this docstring might as well not exist for all the good it'll do me.

File: gutsy-gamblers/main.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    # mobile oriented settings system
+
+    def build_config(self, config):
+        config.setdefaults('global', {
+            'location_friendly': '',
+            'location_latlon': '',
+        })
+
+    def build(self):
+        Builder.load_file('main.kv')
```
Comment (lemonsaurus) [inline gutsy-gamblers/main.py]: nice! using .kv files is :ok_hand:

**Comment by inclement** [comment]
# Summary

A nice complete application combining a number of different features, but with signs of getting a little sloppy during rapid development.

# Version control

Commit quality isn't all that great. Commits mostly have random messages, some not relating to the code at all, although some attempt to describe the change being made. However, it does look like 4 different people were contributing in a consistent way.

# General content and code quality

Overall, looks like a fairly well structured project that neatly combines a number of different features. Things seem sensibly divided into modules and modules and relatively small functions, although there are some parts of the code that seem poorly structured - at least one of which is noted in a comment so at least you recognised it!

The code is in general good, idiomatic Python, with a consistent style throughout. However, there are quite a few sloppy areas that look like they probably arise from time pressure - things like the ugly logic in some places, some of the apparently-arbitrary splitting of content between Python and kv, some code duplication, and commented out code where things have been quickly changed while testing without ever getting polished. Things also aren't totally consistent - notably datahelpers.py is the first and only file with a :param:-formatted docstring, and I think also with type hints.

There are some things that more obviously are worth some attention. I've commented on some examples of questionable code structure, often involving confusing or unclear logic. This kind of thing merits at least some comments, if you think it's the right way to do things despite the issues (which is sometimes fine). There are also quite a few places where the logic seems unnecessarily extended/convoluted, which I've also commented where I saw them.

This leads nicely into the general point that documentation is lacking. It's good that you've put docstrings in some places, but they're clearly the exception rather than the norm. The code flow is quite complex, with information passing between different widgets updated/added/removed in different places, and good documentation can really help to make it clear how this is working. As it is, I suspect some of it is a bit ad hoc and convoluted, probably as a result of the time pressure I guessed about above.

All that said, I don't want to give the impression that the code is very bad, the general flow is fine and much of the questionable structure is looks like an understandable result of rapid learning and development. I suspect you're aware of a lot of it, and would have cleaned it up with more time (...or wouldn't because nobody ever gets around to that). We're still judging on code quality so that isn't a free pass, but I recognise the trade-off in order to produce the large amount of code in the app.

# Use of Kivy

The code shows a good understanding of Kivy. Different gui components are used appropriately to produce a nice gui with various different behaviours. Various different features and tools are combined well, from the use of the EffectWidget (a relatively advanced element for newer users), to appropriate use of canvas instructions to manipulate the primary gui elements, to the overall multi-screen app structure with popups and integration with the location finder.

I do think the balance of code between kv and Python is questionable in places, which is the subject of some of my specific comments. This is a difficult balance, there's nothing technically wrong with what you've done but without a clearly defined structure it can be hard to know when to look in the kv for a behaviour vs when to look in the Python. Better documentation can help, of course, but perhaps also if you went at it again you'd have a stronger idea of how to split things from the start. This is something that can quickly become convoluted in a rapid development situation where you add things when and where they work.

I also notice that your app doesn't seem to apply any style on top o fthe Kivy default, e.g. the Button images are not changed. Choosing some more thematic alternatives might be more consistent for your app.

# README

The summary is nicely written, but I'm left with no idea what the app actually does or how to interact with it. I actually hit some issues running it as well, so I'm reviewing here without some context - all the more ereason to give a clearer summary in the README, even if you keep the funny text as well :) !

A pipenv environment is included, but this isn't mentioned in the README.

On balance it's a good thing that you left in the license details within this repo, but better to rewrite to be specific to your application.

I don't understand the message to the judges :<

**Comment by tshirtman** [comment]
@inclement it seems you commented instead of validating your review, so your in code comments are not visible :grimacing:

**Pull request merged**